codec-mqtt/src/main/java/io/netty/handler/codec/mqtt/MqttDecoder.java

```diff
@@ -57,68 +59,90 @@
     private Object variableHeader;
     private int bytesRemainingInVariablePart;
 
+    private AtomicReference<MqttVersion> mqttVersionRef;
+
+    private MqttVersion mqttVersion() {
+        if (mqttVersionRef == null) {
+            return MqttVersion.MQTT_3_1_1;
+        } else {
+            return mqttVersionRef.get();
+        }
+    }
+
     private final int maxBytesInMessage;
 
     public MqttDecoder() {
-      this(DEFAULT_MAX_BYTES_IN_MESSAGE);
+        this(DEFAULT_MAX_BYTES_IN_MESSAGE);
     }
 
     public MqttDecoder(int maxBytesInMessage) {
+        this(maxBytesInMessage, null);
+    }
+
+    public MqttDecoder(AtomicReference<MqttVersion> mqttVersionRef) {
+        this(DEFAULT_MAX_BYTES_IN_MESSAGE, mqttVersionRef);
+    }
+
+    public MqttDecoder(int maxBytesInMessage, AtomicReference<MqttVersion> mqttVersionRef) {
         super(DecoderState.READ_FIXED_HEADER);
         this.maxBytesInMessage = maxBytesInMessage;
+        this.mqttVersionRef = mqttVersionRef;
     }
 
     @Override
     protected void decode(ChannelHandlerContext ctx, ByteBuf buffer, List<Object> out) throws Exception {
         switch (state()) {
-            case READ_FIXED_HEADER: try {
```

Comment:
It would be nice if you could revert the formatting changes to make this easier to review

Reply:
done wherever possible